distributed/comm/mpi.py

```diff
@@ -0,0 +1,781 @@
+"""
+This is an MVAPICH2-based (http://mvapich.cse.ohio-state.edu) communication 
+backend for the Dask Distributed library.
+
+This is based on the UCX device. UCX: https://github.com/openucx/ucx
+"""
+
+import logging
+import struct
+import weakref
+
+import collections
+import dask
+import asyncio
+import socket
+
+import random
+import time
+import sys
+
+import rmm
+
+from .addressing import parse_host_port, unparse_host_port
+from .core import Comm, Connector, Listener, CommClosedError
+from .registry import Backend, backends
+from .utils import ensure_concrete_host, to_frames, from_frames
+from ..utils import (
+    ensure_ip,
+    get_ip,
+    get_ipv6,
+    nbytes,
+    log_errors,
+    CancelledError,
+    parse_bytes,
+)
+
+from mpi4py import MPI
+
+logger = logging.getLogger(__name__)
+
+host_array = None
+device_array = None
+INITIAL_TAG_OFFSET = 100
+TAG_QUOTA_PER_CONNECTION = 200
+
+# workers and the scheduler randomly select ports for listening to 
+# incoming connections between this range
+LOWER_PORT_RANGE=30000 
+UPPER_PORT_RANGE=40000
+
+# This constant is used to split large message into chunks of this size
+CHUNK_SIZE = 2 ** 30 
+
+initialized = False
+
+# The tag_table dictionary is used to maintain 'tag offset' for a pair of 
+# processes. Imagine a Dask program executing with 4 MPI processes - a scheduler,
+# a client, and 2 workers. The MPI.COMM_WORLD communicator is used by all for
+# exchanging messages. However, during the connection establishment, all processes
+# need to make sure that they have a unique tag range to use with MPI.COMM_WORLD
+# for a pair of processes. Otherwise there is a possibility of message interference 
+# because processes connect with one another dynamically and also there are 
+# control and data channels. This situation is avoided using this dictionary
+# of tags.
+tag_table = dict()
+
+def synchronize_stream(stream=0):
+    import numba.cuda
+
+    ctx = numba.cuda.current_context()
+    cu_stream = numba.cuda.driver.drvapi.cu_stream(stream)
+    stream = numba.cuda.driver.Stream(ctx, cu_stream, None)
+    stream.synchronize()
```

Comment:
Does MPI need to synchronize on the stream before sending? Many things with RMM and cuDF are stream-ordered. So this might not be needed if MPI respects that

Reply:
We are investigating and will update here.